src/permutations.rs

```diff
@@ -153,11 +137,29 @@ where
 
     fn size_hint(&self) -> (usize, Option<usize>) {
         match self.state {
-            PermutationState::StartUnknownLen { .. } |
-            PermutationState::OngoingUnknownLen { .. } => (0, None), // TODO can we improve this lower bound?
+            PermutationState::StartUnknownLen { k } => {
+                let (mut low, mut upp) = self.vals.size_hint();
+                low = CompleteState::Start { n: low, k }.remaining().unwrap_or(usize::MAX);
+                upp = upp.and_then(|n| CompleteState::Start { n, k }.remaining());
+                (low, upp)
+            }
+            PermutationState::OngoingUnknownLen { k, min_n } => {
+                let prev_iteration_count = min_n - k + 1;
+                let (mut low, mut upp) = self.vals.size_hint();
+                low = CompleteState::Start { n: low, k }
+                    .remaining()
+                    .unwrap_or(usize::MAX)
+                    .saturating_sub(prev_iteration_count);
+                upp = upp.and_then(|n| {
+                    CompleteState::Start { n, k }
+                        .remaining()
+                        .map(|count| count.saturating_sub(prev_iteration_count))
+                });
+                (low, upp)
+            }
             PermutationState::Complete(ref state) => match state.remaining() {
-                CompleteStateRemaining::Known(count) => (count, Some(count)),
-                CompleteStateRemaining::Overflow => (::std::usize::MAX, None)
+                Some(count) => (count, Some(count)),
+                None => (::std::usize::MAX, None)
             }
             PermutationState::Empty => (0, Some(0))
         }
```

Comment:
Your individual commit messages have helpful commentary about each of the branches here. Could you include some of those explanations as comments?

Reply:
Done in https://github.com/rust-itertools/itertools/pull/739/commits/95536d6413cb340a7c0c90ca5b3925bfa198ba24